### ISSUE TITLE
Low speed on AMD GPU

### ISSUE THREAD

**Issue by ArkhamInsanity** [city96/ComfyUI-GGUF#48]
Here is my setup, using ubuntu 
AMD 6800 XT 16GB Vram
32GB Ram
Python version: 3.10.12
pytorch version: 2.2.1+rocm5.7

I am getting between 14s-15s/it with flux1-dev-Q2_K.gguf, also Q4_0 and Q6_K, with T5XXL_fp8 and fp16...
 
![image](https://github.com/user-attachments/assets/97f4a039-5c1f-4fc4-8657-34b26c4fac5f)

I only find examples of Nvidia speeds, so I don't know if I should expect faster results or not.

**Comment by city96** [comment]
I'm not sure either what speed to expect as I don't have an AMD card. It looks like it's loading into lowvram mode from your screenshot though. What does your workflow look like?

**Comment by ArkhamInsanity** [comment]
![image](https://github.com/user-attachments/assets/5a1d2c9c-a817-4e8e-af5c-d5a0385c231c)

![image](https://github.com/user-attachments/assets/107f3ea7-3fe0-44e2-b8b6-47989ac21c8b)

I also tried different samplers, including basic workflow from comfyui with custom sampler, same results.

**Comment by mamei16** [comment]
This is around the same speed I get on my 6800M. If you're not doing it already, try using the CLI argument `--force-fp32`

**Comment by fgdfgfthgr-fox** [comment]
> This is around the same speed I get on my 6800M. If you're not doing it already, try using the CLI argument `--force-fp32`

That works for me, I experienced significant speedup after that... but why?

**Comment by ArkhamInsanity** [comment]
Thank you very much mamei16 ! It's much faster this way, I went from 15s/it to 8.73s/it

![image](https://github.com/user-attachments/assets/f8bbf723-ce83-4f1b-ab87-0acfcb91c321)

Same question, why? :D

**Comment by xCentral** [comment]
--force-fp32 should work if I just throw this behind "python3 main.py" in the bash window ?

**Comment by mamei16** [comment]
> > This is around the same speed I get on my 6800M. If you're not doing it already, try using the CLI argument `--force-fp32`
> 
> That works for me, I experienced significant speedup after that... but why?

---

> Thank you very much mamei16 ! It's much faster this way, I went from 15s/it to 8.73s/it
> 
> ![image](https://private-user-images.githubusercontent.com/121449326/359366426-f8bbf723-ce83-4f1b-ab87-0acfcb91c321.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxODAzNjMsIm5iZiI6MTcyNDE4MDA2MywicGF0aCI6Ii8xMjE0NDkzMjYvMzU5MzY2NDI2LWY4YmJmNzIzLWNlODMtNGYxYi1hYjg3LTBhY2ZjYjkxYzMyMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyMFQxODU0MjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMjkyZjc0M2E4ZGExNTIyYzAwNWMzMDI3ZTZjNTQ2MTkwNDdmZGEyYzdiMTM3MzViZTMzOTIwNTFlY2RhOTFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.PVTVBi_Z80qLCUtSKwvUhFATnOJUjqqVaVbufuD6Kfg)
> 
> Same question, why? :D

I have no idea :D

**Comment by mamei16** [comment]
> --force-fp32 should work if I just throw this behind "python3 main.py" in the bash window ?

Yep!

**Comment by AroPix** [comment]
Do you guys use any other launch arguments? I have a 7800 XT and it's still pretty slow for me, even after forcing fp32.

**Comment by mamei16** [comment]
> Do you guys use any other launch arguments? I have a 7800 XT and it's still pretty slow for me, even after forcing fp32. Edit: i get around 8s/it.

I don't know of any other arg that further improves performance. It just seems that the current software does not properly take advantage of AMD GPUs. Even older Nvidia cards perform better.

**Comment by NextTechAndAI** [comment]
> Do you guys use any other launch arguments? I have a 7800 XT and it's still pretty slow for me, even after forcing fp32. Edit: i get around 8s/it.

My 6800 16GB VRAM 32GB RAM has 18.50s/it with DEV FP16, but 10.83s/it with Q5_KS when using ---force-fp32. Most other models have similar performance like FP16 and FP16 even gets much slower by using --force-fp32.
I've tested --highvram, --gpu-only and some more runtime parameters without significant improvements.

**Comment by city96** [comment]
What happens if you change [this line](https://github.com/city96/ComfyUI-GGUF/blob/1ebce77a446407c2a48802111816401051ab9798/dequant.py#L18) in dequant.py from:

```py
        out = dequantize(data, qtype, oshape, dtype=None)
```

To this:
```py
        out = dequantize(data, qtype, oshape, dtype=torch.float32)
```

**Comment by mamei16** [comment]
> What happens if you change [this line](https://github.com/city96/ComfyUI-GGUF/blob/1ebce77a446407c2a48802111816401051ab9798/dequant.py#L18) in dequant.py from:
> 
> ```python
>         out = dequantize(data, qtype, oshape, dtype=None)
> ```
> 
> To this:
> 
> ```python
>         out = dequantize(data, qtype, oshape, dtype=torch.float32)
> ```

Seems to have no effect.

**Comment by NextTechAndAI** [comment]
Same here, no effect after changing dequant.py.
Nevertheless with --force-fp32 there is still a small speed increase for Q8_0 and a 2x speed increase for Q5_KS - with and without changing dequant.py.
I'm on Zluda with ROCm 5.7 and both Q models mentioned are superior to the FP8 model, not least in terms of the quality of the results.

**Comment by city96** [comment]
Interesting, thank you for testing. That change would make the dequantization also run in FP32, but it looks like that's not contributing to the slowdown significantly then.

**Comment by Exploder98** [comment]
> That works for me, I experienced significant speedup after that... but why?

Because by default Comfy tries to use `bfloat16` as the compute datatype for FLUX (ref: https://github.com/comfyanonymous/ComfyUI/blob/7df42b9a2364bae6822fbd9e9fa10cea2e319ba3/comfy/supported_models.py#L645). `bfloat16` seems to be very slow on AMD through pytorch for some reason (no hardware support and therefore it tries to emulate the datatype? idk)

**Comment by NextTechAndAI** [comment]
Thanks for explaining that change and the reference to bfloat16.
BTW, there is no improvement in performance at all when using --force-fp16.
I wanted to see if only a few models profit from --force-fp32:

With --force-fp32:
Q2_K:    9.12s/it
Q3_K_S:  9.59s/it
Q4_0:    9.03s/it
Q4_1:    9.03s/it
Q4_K_S:  9.17s/it
Q5_0:   11.04s/it
Q5_1:   14.29s/it
Q5_K_S: 11.89s/it
Q6_K:   16.17s/it
Q8_0:   20.05s/it

Without any runtime parameters:
Q2_K:   17.96s/it
Q3_K_S: 18.46s/it
Q4_0:   17.90s/it
Q4_1:   17.89s/it
Q4_K_S: 18.04s/it
Q5_0:   18.58s/it
Q5_1:   18.55s/it
Q5_K_S: 18.33s/it
Q6_K:   18.24s/it
Q8_0:   18.64s/it

All tests with SCHNELL 1024x1024, t5xxl_fp16, comparable results with 2 quick DEV tests.
Original FP16 similar to Q8_0.
It seems that with --force-fp32 you sacrifice quality for speed and space - as I would expect from quantized models. For some reason this is not the case without this parameter. Would be interesting to know how this parameter influences other AMD GPUs and NVIDIA GPUs.

**Comment by Exploder98** [comment]
# (Nearly) doubling the inference speed by setting some environment variables

I managed to get the speed (FLUX dev, 1024x1024, Q6_K, t5xxl_fp8_e4m3fn) on my RX 6900 XT from 7.3 s/it to around **3.85 s/it** by running optimization through [PyTorch TunableOp](https://rocm.blogs.amd.com/artificial-intelligence/pytorch-tunableop/README.html). This is still slower than an RTX 3080 (which gets about 2 s/it) which should have comparable hardware performance, but the improvement is nonetheless sizable.

## How I did this

This was actually pretty simple once I found out about TunableOp. It requires at least PyTorch v2.3. To use it, I just had to run ComfyUI like this:

```
PYTORCH_TUNABLEOP_ENABLED=1 PYTORCH_TUNABLEOP_VERBOSE=1 python main.py --novram
```

The initial optimization run will take like half an hour because TunableOp goes through all variants of the matrix multiplication operations trying to find the fastest one for each individual multiplication, but subsequent runs should be fast because the results are saved to a file (`tunableop_results.csv` by default, can be changed with the `PYTORCH_TUNABLEOP_FILENAME` environment variable). I had to run ComfyUI with `--novram` because otherwise I would run out of GPU memory at one point during the optimization process for some reason. When rerunning ComfyUI without this flag afterwards, optimization for one operation did repeat once for some reason (but without OOM). 

## Additional notes

Unfortunately, changing image size causes the optimization to be run again (matrix sizes change), but I think the results are saved incrementally so that going through all needed resolutions once should be enough. Also, updating anything related to PyTorch/ROCm will cause the optimization results to be invalid and therefore the optimization needs to be run again. I don't know if this optimization works on Windows, as I'm running Linux.

Additionally, I have edited [the line I mentioned](https://github.com/comfyanonymous/ComfyUI/blob/7df42b9a2364bae6822fbd9e9fa10cea2e319ba3/comfy/supported_models.py#L645) in my previous comment from

```
supported_inference_dtypes = [torch.bfloat16, torch.float16, torch.float32]
```

to

```
supported_inference_dtypes = [torch.float16, torch.float32]
```

so that ComfyUI would run the inference in fp16. I did not check whether just using `--force-fp32` and TunableOp would bring the same performance boost, but using fp16 should anyway be faster. I also tested the non-GGUF fp8 version, and it is slightly faster compared to GGUF, at around 3.3 s/it.

## Final words

So yeah, it seems that the low performance is caused at least partly by PyTorch choosing non-optimal matrix multiplication operations on AMD GPUs. There's probably still something wonky going on as I think my card should still be a bit faster, but almost doubling the performance with a couple of environment variables is already pretty good.

I wonder how many other projects would benefit from running TunableOp. It would also be interesting to know whether this effect is only limited to consumer cards, or if this is present even on datacenter-class cards.

**Comment by xCentral** [comment]
> # (Nearly) doubling the inference speed by setting some environment variables
> I managed to get the speed (FLUX dev, 1024x1024, Q6_K, t5xxl_fp8_e4m3fn) on my RX 6900 XT from 7.3 s/it to around **3.85 s/it** by running optimization through [PyTorch TunableOp](https://rocm.blogs.amd.com/artificial-intelligence/pytorch-tunableop/README.html). This is still slower than an RTX 3080 (which gets about 2 s/it) which should have comparable hardware performance, but the improvement is nonetheless sizable.
> 
> ## How I did this
> This was actually pretty simple once I found out about TunableOp. It requires at least PyTorch v2.3. To use it, I just had to run ComfyUI like this:
> 
> ```
> PYTORCH_TUNABLEOP_ENABLED=1 PYTORCH_TUNABLEOP_VERBOSE=1 python main.py --novram
> ```
> 
> The initial optimization run will take like half an hour because TunableOp goes through all variants of the matrix multiplication operations trying to find the fastest one for each individual multiplication, but subsequent runs should be fast because the results are saved to a file (`tunableop_results.csv` by default, can be changed with the `PYTORCH_TUNABLEOP_FILENAME` environment variable). I had to run ComfyUI with `--novram` because otherwise I would run out of GPU memory at one point during the optimization process for some reason. When rerunning ComfyUI without this flag afterwards, optimization for one operation did repeat once for some reason (but without OOM).
> 
> ## Additional notes
> Unfortunately, changing image size causes the optimization to be run again (matrix sizes change), but I think the results are saved incrementally so that going through all needed resolutions once should be enough. Also, updating anything related to PyTorch/ROCm will cause the optimization results to be invalid and therefore the optimization needs to be run again. I don't know if this optimization works on Windows, as I'm running Linux.
> 
> Additionally, I have edited [the line I mentioned](https://github.com/comfyanonymous/ComfyUI/blob/7df42b9a2364bae6822fbd9e9fa10cea2e319ba3/comfy/supported_models.py#L645) in my previous comment from
> 
> ```
> supported_inference_dtypes = [torch.bfloat16, torch.float16, torch.float32]
> ```
> 
> to
> 
> ```
> supported_inference_dtypes = [torch.float16, torch.float32]
> ```
> 
> so that ComfyUI would run the inference in fp16. I did not check whether just using `--force-fp32` and TunableOp would bring the same performance boost, but using fp16 should anyway be faster. I also tested the non-GGUF fp8 version, and it is slightly faster compared to GGUF, at around 3.3 s/it.
> 
> ## Final words
> So yeah, it seems that the low performance is caused at least partly by PyTorch choosing non-optimal matrix multiplication operations on AMD GPUs. There's probably still something wonky going on as I think my card should still be a bit faster, but almost doubling the performance with a couple of environment variables is already pretty good.
> 
> I wonder how many other projects would benefit from running TunableOp. It would also be interesting to know whether this effect is only limited to consumer cards, or if this is present even on datacenter-class cards.

Not sure if you're already using it but you can also try adding in flash attention https://github.com/Beinsezii/comfyui-amd-go-fast

**Comment by Exploder98** [comment]
I have not tried that! However, looks like it only works on Navi 3 for now, I have an RX 6900 XT which is a Navi 2 card :(

I think I tried to compile Flash Attention for ROCm at some point and it failed so I don't have high hopes of getting it to work this time...

**Comment by xCentral** [comment]
Sorry to hear. I actually went from 46 seconds to 36 seconds a generation using it with Flux Dev, 24 steps, 1280x720

**Comment by Exploder98** [comment]
> Did you happen to get a HIPBLASLT_TENSILE error when running pytorch tuneable?

I got some warning related to that, but the optimization still worked. Apparently it's possible to disable hipblaslt stuff by setting `PYTORCH_TUNABLEOP_HIPBLASLT_ENABLED=0` (all env vars are listed in [the TunableOp README](https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/cuda/tunable#environment-variable-interface)).

**Comment by xCentral** [comment]
Thank you! So with the tuning parameter, it took off about 3 seconds. So I'm down to 32.9 seconds on average with 24 steps @ 1280x720

**Comment by AroPix** [comment]
Thanks guys! Was able to get it from 8s/it to 2s/it on my 7800 XT using these flags and flash attention. A lot better now! :)

**Comment by ArkhamInsanity** [comment]
> > Did you happen to get a HIPBLASLT_TENSILE error when running pytorch tuneable?
> 
> I got some warning related to that, but the optimization still worked. Apparently it's possible to disable hipblaslt stuff by setting `PYTORCH_TUNABLEOP_HIPBLASLT_ENABLED=0` (all env vars are listed in [the TunableOp README](https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/cuda/tunable#environment-variable-interface)).

I tried it but I get this error 

reading tuning results from tunableop_results0.csv
key="PT_VERSION" is not provided for validation. 
results validator check failed
Memory access fault by GPU node-1 (Agent handle: 0x2f900b0) on address 0x74e206205000. Reason: Page not present or supervisor privilege.
Abandon (core dumped)

**Comment by Exploder98** [comment]
> key="PT_VERSION" is not provided for validation.

This sounds almost like the csv file (`tunableop_results0.csv`) is somehow corrupted. Can you check what its contents look like? Maybe try deleting/renaming the csv file and see if rerunning the optimization works.

**Comment by Exploder98** [comment]
I _think_ CPU RAM usage was around 20 GiB at max, so it should fit in 32 GiB (I have 64). Did you try with `--force-fp32` or did you use my edit to supported_models.py? I tried with `--force-fp32` and got some OOMs with that. Also, increasing the batch size even from 1 to 2 seems to cause OOMs during tuning. Remember to use `--no-vram` for the tuning run, otherwise you'll get an OOM.

I did not try with LoRAs though. Maybe doing the tuning once first without LoRA and then adding the LoRA afterwards could work?

**Comment by harakiru** [comment]
> I _think_ CPU RAM usage was around 20 GiB at max, so it should fit in 32 GiB (I have 64). Did you try with `--force-fp32` or did you use my edit to supported_models.py? I tried with `--force-fp32` and got some OOMs with that. Also, increasing the batch size even from 1 to 2 seems to cause OOMs during tuning. Remember to use `--no-vram` for the tuning run, otherwise you'll get an OOM.
> 
> I did not try with LoRAs though. Maybe doing the tuning once first without LoRA and then adding the LoRA afterwards could work?

I tried doing it for sdxl model and after updating rocm+torch to the latest version the OOM issue went away. I'm now struggling with the .csv file not being generated correctly. It creates a .csv file when i first run it, however it stays empty after tuning, and when i try to run it again it complains about "reading tuning results from tunableop_results0.csv key="PT_VERSION" is not provided for validation. results validator check failed". I dont really really know what else might be the problem since it doesn't spit out any error messages aside from that.

**Comment by fgdfgfthgr-fox** [comment]
> I tried doing it for sdxl model and after updating rocm+torch to the latest version the OOM issue went away. I'm now struggling with the .csv file not being generated correctly. It creates a .csv file when i first run it, however it stays empty after tuning, and when i try to run it again it complains about "reading tuning results from tunableop_results0.csv key="PT_VERSION" is not provided for validation. results validator check failed". I dont really really know what else might be the problem since it doesn't spit out any error messages aside from that.

Use ctrl+c on the terminal to turn the comfyui off and you should see a message saying it's writing the result to the file before the terminal closes.

**Comment by ArkhamInsanity** [comment]
> Rocm 6's focus has been to optimize fp8 performance on Navi 3 from the documentation I can find. Might be worth it to try this model with a 6.1 Rocm Linux setup https://civitai.com/models/637170?modelVersionId=724032 I feel you should get better performance than going GGUF. The model is just slightly under 16G, so you'll need to test with your 7800. Works incredibly well on my 7900xtx

Yep I am trying it today, and I am getting 7.18s/it with fp8 flux1 diffusion model and t5xxl_fp8. Need to try different setups (GGUF flux + fp8 t5xxl, fp8 flux1 + GGUF t5xxl), to see how it goes

**Comment by harakiru** [comment]
> > I tried doing it for sdxl model and after updating rocm+torch to the latest version the OOM issue went away. I'm now struggling with the .csv file not being generated correctly. It creates a .csv file when i first run it, however it stays empty after tuning, and when i try to run it again it complains about "reading tuning results from tunableop_results0.csv key="PT_VERSION" is not provided for validation. results validator check failed". I dont really really know what else might be the problem since it doesn't spit out any error messages aside from that.
> 
> Use ctrl+c on the terminal to turn the comfyui off and you should see a message saying it's writing the result to the file before the terminal closes.

Figured it out. It was because i was launching it through comfyui-manager's comfy-cli. You need to launch comfyui through "python main.py" or it wont write the .csv file. Will tinker around with it more when i get the chance.

**Comment by cyber827** [comment]
> > > I tried doing it for sdxl model and after updating rocm+torch to the latest version the OOM issue went away. I'm now struggling with the .csv file not being generated correctly. It creates a .csv file when i first run it, however it stays empty after tuning, and when i try to run it again it complains about "reading tuning results from tunableop_results0.csv key="PT_VERSION" is not provided for validation. results validator check failed". I dont really really know what else might be the problem since it doesn't spit out any error messages aside from that.
> > 
> > 
> > Use ctrl+c on the terminal to turn the comfyui off and you should see a message saying it's writing the result to the file before the terminal closes.
> 
> Figured it out. It was because i was launching it through comfyui-manager's comfy-cli. You need to launch comfyui through "python main.py" or it wont write the .csv file. Will tinker around with it more when i get the chance.

Same issue on Windows, the csv file remains empty, getting the PT_VERSION error, running from main.py does not work. Any idea on how to fix it? What exactly is generated in the first row of the csv file?

**Comment by DeeH0pe** [comment]
> > > > I tried doing it for sdxl model and after updating rocm+torch to the latest version the OOM issue went away. I'm now struggling with the .csv file not being generated correctly. It creates a .csv file when i first run it, however it stays empty after tuning, and when i try to run it again it complains about "reading tuning results from tunableop_results0.csv key="PT_VERSION" is not provided for validation. results validator check failed". I dont really really know what else might be the problem since it doesn't spit out any error messages aside from that.
> > > 
> > > 
> > > Use ctrl+c on the terminal to turn the comfyui off and you should see a message saying it's writing the result to the file before the terminal closes.
> > 
> > 
> > Figured it out. It was because i was launching it through comfyui-manager's comfy-cli. You need to launch comfyui through "python main.py" or it wont write the .csv file. Will tinker around with it more when i get the chance.
> 
> Same issue on Windows, the csv file remains empty, getting the PT_VERSION error, running from main.py does not work. Any idea on how to fix it? What exactly is generated in the first row of the csv file?

Try with --force-fp16. Did it for me

**Comment by DeeH0pe** [comment]
> > key="PT_VERSION" is not provided for validation.
> 
> This sounds almost like the csv file (`tunableop_results0.csv`) is somehow corrupted. Can you check what its contents look like? Maybe try deleting/renaming the csv file and see if rerunning the optimization works.

But how can I make comfy load the file after the "tuning"?

**Comment by alessandroberna** [comment]
> Same issue on Windows, the csv file remains empty, getting the PT_VERSION error, running from main.py does not work. Any idea on how to fix it? What exactly is generated in the first row of the csv file?

I launched ComfyUI with this command: `PYTORCH_TUNABLEOP_ENABLED=1 PYTORCH_TUNABLEOP_VERBOSE=1 PYTORCH_TUNABLEOP_HIPBLASLT_ENABLED=0 python3 main.py --force-fp32 --novram` (under Linux)
this is the resulting `tunableop_results0.csv` for my 6700XT.
```
Validator,PT_VERSION,2.4.1
Validator,ROCBLAS_VERSION,4.1.0-cefa4a9b-dirty
Validator,ROCM_VERSION,6.1.0.0-82-5fabb4c
Validator,GCN_ARCH_NAME,gfx1030
GemmTunableOp_float_NT,nt_16384_16384_512,Default,21.33
GemmStridedBatchedTunableOp_float_NN,nn_128_256_4352_B_24,Gemm_Rocblas_18717,0.781374
GemmStridedBatchedTunableOp_float_NN,nn_128_4096_4352_B_24,Gemm_Rocblas_18718,10.5327
GemmStridedBatchedTunableOp_float_TN,tn_4352_256_128_B_24,Gemm_Rocblas_19997,0.718891
GemmStridedBatchedTunableOp_float_TN,tn_4352_4096_128_B_24,Gemm_Rocblas_19997,10.902
GemmTunableOp_float_TN,tn_16384_512_16384,Gemm_Rocblas_19999,25.1117
GemmTunableOp_float_TN,tn_6144_1_3072,Gemm_Rocblas_20194,0.212292
GemmTunableOp_float_TN,tn_3072_4352_15360,Default,39.2634
GemmTunableOp_float_TN,tn_9216_1_3072,Gemm_Rocblas_20192,0.403658
GemmTunableOp_float_TN,tn_3072_256_12288,Gemm_Rocblas_20006,2.09865
GemmTunableOp_float_TN,tn_3072_4096_64,Gemm_Rocblas_19950,0.167817
GemmTunableOp_float_TN,tn_9216_4096_3072,Gemm_Rocblas_19980,21.0174
GemmTunableOp_float_TN,tn_3072_1_256,Gemm_Rocblas_20190,0.0169701
GemmTunableOp_float_TN,tn_3072_1_3072,Gemm_Rocblas_20190,0.12163
GemmTunableOp_float_TN,tn_64_4096_3072,Gemm_Rocblas_20112,0.253881
GemmTunableOp_float_TN,tn_3072_1_768,Gemm_Rocblas_20190,0.0390971
GemmTunableOp_float_TN,tn_12288_256_3072,Default,1.85803
GemmTunableOp_float_TN,tn_18432_1_3072,Gemm_Rocblas_20142,0.903636
GemmTunableOp_float_TN,tn_9216_256_3072,Gemm_Rocblas_19999,1.46172
GemmTunableOp_float_TN,tn_3072_4096_12288,Gemm_Rocblas_19997,29.1898
GemmTunableOp_float_TN,tn_12288_4096_3072,Gemm_Rocblas_19980,28.2916
GemmTunableOp_float_TN,tn_21504_4352_3072,Gemm_Rocblas_19980,52.7649
GemmTunableOp_float_TN,tn_3072_256_4096,Gemm_Rocblas_20006,0.637689
GemmTunableOp_float_TN,tn_3072_4096_3072,Gemm_Rocblas_19963,7.14236
GemmTunableOp_float_TN,tn_3072_256_3072,Gemm_Rocblas_20006,0.524814
```
I had the same `"reading tuning results from tunableop_results0.csv key="PT_VERSION" is not provided for validation. results validator check failed".` warning as i had an empty .csv from a previous incomplete run.  
The .csv remained empty during inferencing and even after a sample image had been generated, it was populated only after shutting down ComfyUI like others have reported.

**Comment by cyber827** [comment]
> > Same issue on Windows, the csv file remains empty, getting the PT_VERSION error, running from main.py does not work. Any idea on how to fix it? What exactly is generated in the first row of the csv file?
> 
> I launched ComfyUI with this command: `PYTORCH_TUNABLEOP_ENABLED=1 PYTORCH_TUNABLEOP_VERBOSE=1 PYTORCH_TUNABLEOP_HIPBLASLT_ENABLED=0 python3 main.py --force-fp32 --novram` (under Linux) this is the resulting `tunableop_results0.csv` for my 6700XT.
> 
> ```
> Validator,PT_VERSION,2.4.1
> Validator,ROCBLAS_VERSION,4.1.0-cefa4a9b-dirty
> Validator,ROCM_VERSION,6.1.0.0-82-5fabb4c
> Validator,GCN_ARCH_NAME,gfx1030
> GemmTunableOp_float_NT,nt_16384_16384_512,Default,21.33
> GemmStridedBatchedTunableOp_float_NN,nn_128_256_4352_B_24,Gemm_Rocblas_18717,0.781374
> GemmStridedBatchedTunableOp_float_NN,nn_128_4096_4352_B_24,Gemm_Rocblas_18718,10.5327
> GemmStridedBatchedTunableOp_float_TN,tn_4352_256_128_B_24,Gemm_Rocblas_19997,0.718891
> GemmStridedBatchedTunableOp_float_TN,tn_4352_4096_128_B_24,Gemm_Rocblas_19997,10.902
> GemmTunableOp_float_TN,tn_16384_512_16384,Gemm_Rocblas_19999,25.1117
> GemmTunableOp_float_TN,tn_6144_1_3072,Gemm_Rocblas_20194,0.212292
> GemmTunableOp_float_TN,tn_3072_4352_15360,Default,39.2634
> GemmTunableOp_float_TN,tn_9216_1_3072,Gemm_Rocblas_20192,0.403658
> GemmTunableOp_float_TN,tn_3072_256_12288,Gemm_Rocblas_20006,2.09865
> GemmTunableOp_float_TN,tn_3072_4096_64,Gemm_Rocblas_19950,0.167817
> GemmTunableOp_float_TN,tn_9216_4096_3072,Gemm_Rocblas_19980,21.0174
> GemmTunableOp_float_TN,tn_3072_1_256,Gemm_Rocblas_20190,0.0169701
> GemmTunableOp_float_TN,tn_3072_1_3072,Gemm_Rocblas_20190,0.12163
> GemmTunableOp_float_TN,tn_64_4096_3072,Gemm_Rocblas_20112,0.253881
> GemmTunableOp_float_TN,tn_3072_1_768,Gemm_Rocblas_20190,0.0390971
> GemmTunableOp_float_TN,tn_12288_256_3072,Default,1.85803
> GemmTunableOp_float_TN,tn_18432_1_3072,Gemm_Rocblas_20142,0.903636
> GemmTunableOp_float_TN,tn_9216_256_3072,Gemm_Rocblas_19999,1.46172
> GemmTunableOp_float_TN,tn_3072_4096_12288,Gemm_Rocblas_19997,29.1898
> GemmTunableOp_float_TN,tn_12288_4096_3072,Gemm_Rocblas_19980,28.2916
> GemmTunableOp_float_TN,tn_21504_4352_3072,Gemm_Rocblas_19980,52.7649
> GemmTunableOp_float_TN,tn_3072_256_4096,Gemm_Rocblas_20006,0.637689
> GemmTunableOp_float_TN,tn_3072_4096_3072,Gemm_Rocblas_19963,7.14236
> GemmTunableOp_float_TN,tn_3072_256_3072,Gemm_Rocblas_20006,0.524814
> ```
> 
> I had the same `"reading tuning results from tunableop_results0.csv key="PT_VERSION" is not provided for validation. results validator check failed".` warning as i had an empty .csv from a previous incomplete run. The .csv remained empty during inferencing and even after a sample image had been generated, it was populated only after shutting down ComfyUI like others have reported.
> 
> Edit: after removing the flag `--novram` it started giving me `Memory access fault by GPU node-1 (Agent handle: 0x2f900b0) on address 0x74e206205000. Reason: Page not present or supervisor privilege.` and a segmentation fault, even if part of the initial optimization was already done
> 
> i fixed this by editing `supported_inference_dtypes` in `ComfyUI/comfy/supported_models.py`and removing the `--force-fp32` flag as Exploder98 suggested. Some more entries have been added to my .csv, i am now getting 13.65s/it with flux1-dev-Q5_K_S With --force-fp32 i was getting around 25s/it, stock was giving me 40-45 s/it

Thanks for the info, I got it working but did not notice any improvement. I'm getting around 9s/it with the same card but on zluda and Q8.

**Comment by zezic** [comment]
As long as this thread became something heavily related to getting better speed with AMD GPUs x ComfyUI x FLUX, I allow myself to share this block caching technique with you: https://github.com/chengzeyi/Comfy-WaveSpeed

It affects the image, but it improves the speed heavily in some scenarios. However, I'm not sure that `more steps` + `fisrt block cache` is better than just `less steps`, have to experiment with this.

**Comment by DeeH0pe** [comment]
> As long as this thread became something heavily related to getting better speed with AMD GPUs x ComfyUI x FLUX, I allow myself to share this block caching technique with you: https://github.com/chengzeyi/Comfy-WaveSpeed
> 
> It affects the image, but it improves the speed heavily in some scenarios. However, I'm not sure that `more steps` + `fisrt block cache` is better than just `less steps`, have to experiment with this.

Is this working on windows or linux only?

**Comment by Charmandrigo** [comment]
I can't seem to make it work... at least on Rocm 6.3 over WSL.
For reference I'm using Kohya for training models:
running
```
export PYTORCH_TUNABLEOP_ENABLED=0
export PYTORCH_TUNABLEOP_TUNING=0
export PYTORCH_TUNABLEOP_VERBOSE=0
```

I did a matmul.py test and let it apply tunings.

With TunableOP, the training speed turns into over 60s/t to over 100.
if it's disabled, it can go as fast as 2.7s/it on batches of 8

tbh I wish I knew why this feature makes the speed of operations amusingly slow

**Comment by zezic** [comment]
> Is this working on windows or linux only?

I tested it only on Linux, but after comparisons made by eye along few days I decided to stop using it because it didn't improve speed without sacrificing image quality. So, no magic, unfortunately. Maybe it's more useful for videos, not for stills, I'm not sure.

**Comment by HannesStrohkopp** [comment]
> Here is my setup, using ubuntu AMD 6800 XT 16GB Vram 32GB Ram Python version: 3.10.12 pytorch version: 2.2.1+rocm5.7
> 
> I am getting between 14s-15s/it with flux1-dev-Q2_K.gguf, also Q4_0 and Q6_K, with T5XXL_fp8 and fp16...
> 
> ![image](https://github.com/user-attachments/assets/97f4a039-5c1f-4fc4-8657-34b26c4fac5f)
> 
> I only find examples of Nvidia speeds, so I don't know if I should expect faster results or not.

Would you mind sharing how exactly you got it to work? Cant get the 6800XT to work, as it is unsupported by ROCm...

**Comment by Holmes2002** [comment]
any solution, I tried GGUF flux 4b However speed was very slow, about 7s/step -> ~2m for an image
